### PR TITLE
Add tutorials for auxiliary and connector implementation

### PR DESCRIPTION
**Pull request by sebclrsn**
Added files:
- how_to_connector.rst : how to implement a generic connector and flasher
- how_to_auxiliary.rst : which interfaces for which purpose, Thread auxiliary implementation example

Changed files:
- lib API references added in api.rst
- sphinx warnings resolved

### CODE REVIEW

File: docs/how_to_auxiliary.rst [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+
+Each command execution is handled in a thread-safe way by getting values from an input queue and
+returning the command result in an output queue.
+
+Each time the execution is entered, the following actions are performed:
+
+1. Verify if a request is available in the input queue
+
+  #. If the command message is "create_auxiliary_instance" and the auxiliary is not created yet,
+     call the ``_create_auxiliary_instance`` method and put a boolean corresponding the success
```
Comment (Pog3k) [inline docs/how_to_auxiliary.rst:56]:
```suggestion
     call the ``_create_auxiliary_instance`` method and put a boolean corresponding to the success

```

File: docs/how_to_connector.rst [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+
+Thus, its goal is to implement the
+sending and reception of data on the lower level protocol layers
+(e.g. CAN, UART, UDP).
+
+Both can be used as context managers as they inherit the common interface
+:py:class:`~pykiso.connector.Connector`.
+
+.. note::
+  Many already implemented connector are available under :py:mod:`pykiso.lib.connectors`.
```
Comment (Pog3k) [inline docs/how_to_connector.rst:18]:
```suggestion
  Many already implemented connectors are available under :py:mod:`pykiso.lib.connectors`.

```

File: docs/how_to_connector.rst [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+
+
+Interface completion
+^^^^^^^^^^^^^^^^^^^^
+
+If the code above is left as such, it won't be usable as a connector as
+the communication channel's abstract methods aren't implemented.
+
+Therefore, all four methods ``_cc_open``, ``_cc_close``, ``_cc_send`` and
+``_cc_receive`` need to be implemented.
```
Comment (Pog3k) [inline docs/how_to_connector.rst:89]:
```suggestion
``_cc_receive`` needs to be implemented.

```

**Review by Pog3k — APPROVED** [review]
just seem my suggestions

File: docs/how_to_connector.rst (hunk 3 above)
Comment (sebclrsn) [reply docs/how_to_connector.rst:89]: This one was correct ;)

File: docs/how_to_connector.rst (hunk 2 above)
Comment (sebclrsn) [reply docs/how_to_connector.rst]: Fixed

File: docs/how_to_auxiliary.rst (hunk 1 above)
Comment (sebclrsn) [reply docs/how_to_auxiliary.rst]: Fixed

**Review by BKaDamien — APPROVED** [review]

File: docs/how_to_auxiliary.rst [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+The execution of concurrent auxiliaries (i.e. inheriting from :py:class:`~pykiso.interfaces.thread_auxiliary.AuxiliaryInterface`
+or :py:class:`~pykiso.interfaces.mp_auxiliary.MpAuxiliaryInterface`) is
+handled by the interfaces' :py:meth:`run method <pykiso.interfaces.thread_auxiliary.AuxiliaryInterface.run>`.
+
+Each command execution is handled in a thread-safe way by getting values from an input queue and
+returning the command result in an output queue.
+
+Each time the execution is entered, the following actions are performed:
+
+1. Verify if a request is available in the input queue
```
[the remark this answers is not in the document]
Comment (sebclrsn) [reply docs/how_to_auxiliary.rst:56]: As the auxiliary creation and deletion are handled in the auxiliary run, it seemed logical to leave them here to provide more precisions on the creation/deletion process of concurrent auxiliaries (in contrary to the section that concerns any auxilary included the simple one)
Also, if I split this in 2 I'll kind of lose the run_command & abort_command descriptions, or should I then still include then in the Auxiliary's run part?

File: docs/how_to_auxiliary.rst [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+Pykiso provides three different auxiliary interfaces, used as basis for
+any implemented auxiliary. These different interfaces aim to cover every
+possible usage of an auxiliary:
+
+- The :py:class:`~pykiso.interfaces.thread_auxiliary.AuxiliaryInterface`
+  is a `Thread <https://docs.python.org/3.7/library/threading.html#threading.Thread>`_-based auxiliary.
+  It is suited for IO-bound tasks where the reception of data cannot be expected.
+- The :py:class:`~pykiso.interfaces.mp_auxiliary.MpAuxiliaryInterface`
+  is a `Process <https://docs.python.org/3.7/library/multiprocessing.html#multiprocessing.Process>`_-based auxiliary.
+  It is suited for CPU-bound tasks where the reception of data cannot be expected and its processing can be CPU-intensive.
```
Comment (sebastianpfischer) [inline docs/how_to_auxiliary.rst:24]: Can it be extended with the information that it utilizes all the cores available on the CPU?

File: docs/how_to_auxiliary.rst [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+- The :py:class:`~pykiso.interfaces.simple_auxiliary.SimpleAuxiliaryInterface`
+  does not implement any kind of concurrent execution. It is suited for host-based applications where the auxiliary
```
Comment (sebastianpfischer) [inline docs/how_to_auxiliary.rst:26]: i miss somehow the information that there is no task or thread running in the background or that it is a sync(hronous) API (vs an async APIs for the other auxiliaries)

**Review by sebastianpfischer — CHANGES_REQUESTED** [review]

File: docs/how_to_auxiliary.rst [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+  initiates every possible action, i.e. the reception of data can always be expected.
+
+Execution of an Auxiliary
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Auxiliary creation and deletion
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Any auxiliary is **created** at test setup by calling :py:meth:`~pykiso.interfaces.thread_auxiliary.AuxiliaryInterface.create_instance`
```
Comment (sebastianpfischer) [inline docs/how_to_auxiliary.rst:35]: What is exactly meant with "Test setup"? Is it test-suite setup? Or TC setup or even before?

File: docs/how_to_auxiliary.rst [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+
+All of the above described Auxiliary interfaces require the same abstract methods
+to be implemented:
+
+- ``_create_auxiliary_instance``: handle the auxiliary creation. Minimal actions to perform are
+  opening the attached ``CChannel``, to which can be added actions such as flashing the device under test,
+  perform security related operations to allow the communication, etc.
+- ``_delete_auxiliary_instance``: handle the auxiliary deletion. This method is the counterpart of
+  ``_create_auxiliary_instance`` and should at least close the opened ``CChannel``.
+
```
Comment (sebastianpfischer) [inline docs/how_to_auxiliary.rst:93]: I think something like "_delete_auxiliary_instance" needs to be done in a way that "_create_auxiliariy_instance" can be called again without side effects.
is missing

File: docs/how_to_auxiliary.rst [hunk 9]
```diff
@@ -0,0 +1,243 @@
...
+Concurrent auxiliary methods
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+In addition to the previously described methods, the concurrent Auxiliary
+interfaces :py:class:`AuxiliaryInterface <pykiso.interfaces.thread_auxiliary.AuxiliaryInterface>`
+and :py:class:`MpAuxiliaryInterface <pykiso.interfaces.mp_auxiliary.MpAuxiliaryInterface>` require
+the following methods to be implemented:
+
+- ``_run_command``: implement the different commands that should be performed by the Auxiliary.
```
Comment (sebastianpfischer) [inline docs/how_to_auxiliary.rst:102]: I miss the information that the developer can implement/extend his auxiliaries with additional methods and that these methods (or in our terminology, commands) need to be implemented here?

File: docs/how_to_connector.rst [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+will derive from its location and constructor parameters:
+
+.. code:: yaml
+
+    connectors:
+      my_chan:
+        # provide the constructor parameters
+        config:
+          # arg1 and arg2 are mandatory as we defined them as positional arguments
+          arg1: value for positional argument arg1
```
Comment (sebastianpfischer) [inline docs/how_to_connector.rst:69]: I may have read fast, but did we describe somewhere that the arguments of __init__() are linked to the arguments passed here?

**Review by sebastianpfischer — CHANGES_REQUESTED** [review]

File: docs/how_to_auxiliary.rst (hunk 6 above)
Comment (sebclrsn) [reply docs/how_to_auxiliary.rst:26]: That's the definition of concurrency (or concurrent execution), should this still be rephrased?

File: docs/how_to_auxiliary.rst (hunk 7 above)
Comment (sebclrsn) [reply docs/how_to_auxiliary.rst:35]: This is the overall test setup as defined by the pykiso architecture (pykiso.test_setup / pykiso.test_execution)

File: docs/how_to_auxiliary.rst (hunk 9 above)
Comment (sebclrsn) [reply docs/how_to_auxiliary.rst:102]: This is illustrated in the code example, I'll explain it further here if necessary

File: docs/how_to_connector.rst (hunk 10 above)
Comment (sebclrsn) [reply docs/how_to_connector.rst:69]: Yes, the lines above show the constructor, this yaml sample is there to illustrate the config file structure resulting of such a constructor

File: docs/how_to_auxiliary.rst (hunk 9 above)
Comment (sebastianpfischer) [reply docs/how_to_auxiliary.rst:102]: I think it needs to be a bit more explicit here.

Comment (sebclrsn) [reply docs/how_to_auxiliary.rst]: Done!

File: docs/how_to_auxiliary.rst (hunk 7 above)
Comment (sebclrsn) [reply docs/how_to_auxiliary.rst]: I explained these terms a bit further

File: docs/how_to_auxiliary.rst (hunk 5 above)
Comment (sebclrsn) [reply docs/how_to_auxiliary.rst:24]: Done

File: src/pykiso/test_coordinator/test_result.py [hunk 11]
```diff
@@ -1,160 +1,161 @@
...
+# Copyright (c) 2010-2021 Robert Bosch GmbH
+# This program and the accompanying materials are made available under the
+# terms of the Eclipse Public License 2.0 which is available at
+# http://www.eclipse.org/legal/epl-2.0.
+#
+# SPDX-License-Identifier: EPL-2.0
+##########################################################################
+
+"""
+test_result
```
[the remark this answers is not in the document]
Comment (sebclrsn) [reply src/pykiso/test_coordinator/test_result.py]: Reworked it

File: docs/how_to_auxiliary.rst [hunk 12]
```diff
@@ -0,0 +1,240 @@
...
+
+Auxiliary implementation example
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+See below an example implementing the basic functionalities of a Thread Auxiliary:
+
+.. code:: python
+
+    import logging
+    from pykiso import AuxiliaryInterface, MpAuxiliaryInterface, CChannel, Flasher
```
[the remark this answers is not in the document]
Comment (sebclrsn) [reply docs/how_to_auxiliary.rst]: Removed it

**Review by sebastianpfischer — APPROVED** [review]

**Pull request merged**